api/s2n.h

```diff
@@ -596,6 +596,258 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
+ * 
+ * @param config A pointer to the s2n_config object being updated.
+ * @param mode The pre-shared key mode to be set. The supported psk modes are listed in the enum `s2n_psk_mode` above. 
+ */
+S2N_API 
+int s2n_config_set_psk_mode(struct s2n_config *config, s2n_psk_mode mode);
+
+/**
+ * Sets the pre-shared key mode on the s2n connection object.
+ * The supported pre-shared key modes are listed in the enum `s2n_psk_mode` above. 
+ * This API overrides the pre-shared key mode for the connection and will continue to use the existing 
+ * pre-shared key mode even when a new config is set for the connection.
+ * 
+ * @param conn A pointer to the s2n_connection object being updated.
+ * @param mode The pre-shared key mode to be set.
+ */
+S2N_API 
+int s2n_connection_set_psk_mode(struct s2n_connection *conn, s2n_psk_mode mode);
+
+/**
+ * Gets the negotiated pre-shared key identity length from the s2n connection object. 
+ * If the negotiated pre-shared key does not exist, the value `0` is returned.  
+ * 
+ * @param conn A pointer to the s2n_connection object that successfully negotiated a pre-shared key connection.
+ * @param identity_length The length of the negotiated pre-shared key identity. 
+ */
+S2N_API 
+int s2n_connection_get_negotiated_psk_identity_length(struct s2n_connection *conn, uint16_t *identity_length);
+
+/**
+ * Gets the negotiated pre-shared key identity from the s2n connection object. 
+ * If the negotiated pre-shared key does not exist, the PSK identity will not be obtained and no error will be returned. 
+ * 
+ * @param conn A pointer to the s2n_connection object.
+ * @param identity The negotiated pre-shared key identity obtained from the s2n_connection object. 
+ * @param max_identity_length The maximum length for the pre-shared key identity. If the negotiated psk_identity length is 
+ * greater than this `max_identity_length` value an error will be returned.
+ */
+S2N_API 
+int s2n_connection_get_negotiated_psk_identity(struct s2n_connection *conn, uint8_t *identity, uint16_t max_identity_length);
+
+struct s2n_offered_psk;
+
+/**
+ * Creates a new s2n offered pre-shared key object. 
+ * An offered pre-shared key object represents a single pre-shared key (PSK) shared by the client.
+ * 
+ * # Safety
+ * 
+ * Use `s2n_offered_psk_free` to free the memory allocated to the s2n offered pre-shared key object created by this API. 
+ *
+ * @return struct s2n_offered_psk* Returns a pointer to the newly created offered pre-shared key object.
+ */
+S2N_API 
+struct s2n_offered_psk* s2n_offered_psk_new();
+
+/**
+ * Frees the memory associated with the `s2n_offered_psk` object.
+ *
+ * @param psk A pointer to the `s2n_offered_psk` object to be freed.
+ */
+S2N_API 
+int s2n_offered_psk_free(struct s2n_offered_psk **psk);
+
+/**
+ * Gets the pre-shared key identity and pre-shared key identity length for a given offered pre-shared key object. 
+ * 
```

Comment:
What's the lifetime of the values returned from this API? How can I use them safely?

Reply:
Offered PSKs dont have a lifetime  - we don't provide one for it.

Reply:
This API returns a pointer to memory. We need to tell users how to safely use that memory.